ods capability statement

APageNHS opened the pull request:
ODS capability statement, with initial search params, package updated to ensure validation, and valid examples (to UK Core + extensions/codesystems)

APageNHS commented on CapabilityStatement/CapabilityStatement-England-ODS.xml:
```diff
@@ -0,0 +1,145 @@
+        <definition value="http://hl7.org/fhir/SearchParameter/address-country" />
+        <type value="token" />
+      </searchParam>
+      <searchParam>
```
We could probably give ODS this capability statement without these custom searchParams, but these are the specific parts of their datamodel they are going to want to search on

At a minimum, role and lastchange date are probably essential

APageNHS replied:
having checked with the supplier doing ODS fhir server / api implementation, lastchangedate, role, active role are the ones they were expecting at a minimum

APageNHS replied:
removed additional searchparams after further discussion with ods

johngeo reviewed (commented):
Nice work.  I've checked that this capability statement is valid against the HL7.org capabilitystatement.xsd.  Just a question about searchParam
        name value="name" /
        definition value="http://hl7.org/fhir/SearchParameter/name" /
        type value="string" /
      /searchParam  .  What is the reason for the type value="string" rather than type value="token", like in the searchParam before this?

APageNHS commented:
> Nice work. I've checked that this capability statement is valid against the HL7.org capabilitystatement.xsd. Just a question about searchParam name value="name" / definition value="http://hl7.org/fhir/SearchParameter/name" / type value="string" / /searchParam . What is the reason for the type value="string" rather than type value="token", like in the searchParam before this?

I was basing the string off the list of search params here: https://hl7.org/fhir/R4/organization.html, though that does mean address-country is wrong, will update that

johngeo reviewed (commented):
Looking at Kevin Sprague's Capability Statement Guidance document, this is missing CapabilityStatement.title, I appreciate this is optional, but in the guidance doc its mandated for an NHS E IG.  I'm using your Capability Statement, to review Kevin's Capability Statement Guidance document ;-).

digitalhealthmartin reviewed (commented):
- 'publisher' should be 'NHS England Interoperability Standards Team' (based on the guidance, although our other assets in NHSE IG uses 'NHS England' so maybe clarify this? I think 'NHS England' should be suffient but let's confirm

- 'contact' should be 'NHS England Interoperability Standards Team' (based on the guidance, although our other assets in NHSE IG uses 'NHS England' so maybe clarify this? I think 'NHS England' should be suffient but let's confirm

- I suggest that the id, name etc should not have the word 'conformance' in the name but instead have 'requirements' or the type of CapapbilityStatement it is. For instance, id = 'England-ODS-Requirements' etc. I will raise this as part of the team-wide review of his CapabailityStatement guidance doc.

- Update description and purpose to remove 'Conformance' and just mention 'Requirements', related to point above (again needs discussion)

- For 'jurisdiction', the currently system is 'iso:3166:-2'. However, don't think 'GB-ENG' is part of that system. According to https://www.gov.uk/government/publications/open-standards-for-government/country-codes, the ISO to be used is ISO 3166-1 alpha-2. Although need discussion about this. Also need clearer guidance about the system value so will feed this back in the guidance review.

- For implementationGuide value, perhaps remove '?version=current' from value as https://simplifier.net/guide/organisational-data-services will point to the version we want a user to be redirected to

- Probably want to clarify whether OAuth is being used and other security info

- The rest security description seems to go onto new lines without any tags. I think it shouldn't go onto new lines until the end of the line is reached.

- Rest interaction code 'read' is used. However, do we also need 'vread' and 'search'? Will need to look at list of use cases if these are appropriate. From my interpreation of 'read', it looks like it's just to view only one resource, such as by id. 

- Is it needed to add the searchParams using hl7.org? I also can't seem to find these in the base standard but maybe in a place I didn't look. I would have thought we just need searchParams specific to the ODS use-case

- Should the searchParam for ods-lastChangeDate use the type of 'date' instead of 'token'?

APageNHS commented:
> * 'publisher' should be 'NHS England Interoperability Standards Team' (based on the guidance, although our other assets in NHSE IG uses 'NHS England' so maybe clarify this? I think 'NHS England' should be suffient but let's confirm
> * 'contact' should be 'NHS England Interoperability Standards Team' (based on the guidance, although our other assets in NHSE IG uses 'NHS England' so maybe clarify this? I think 'NHS England' should be suffient but let's confirm
> * I suggest that the id, name etc should not have the word 'conformance' in the name but instead have 'requirements' or the type of CapapbilityStatement it is. For instance, id = 'England-ODS-Requirements' etc. I will raise this as part of the team-wide review of his CapabailityStatement guidance doc.
> * Update description and purpose to remove 'Conformance' and just mention 'Requirements', related to point above (again needs discussion)
> * For 'jurisdiction', the currently system is 'iso:3166:-2'. However, don't think 'GB-ENG' is part of that system. According to https://www.gov.uk/government/publications/open-standards-for-government/country-codes, the ISO to be used is ISO 3166-1 alpha-2. Although need discussion about this. Also need clearer guidance about the system value so will feed this back in the guidance review.
> * For implementationGuide value, perhaps remove '?version=current' from value as https://simplifier.net/guide/organisational-data-services will point to the version we want a user to be redirected to
> * Probably want to clarify whether OAuth is being used and other security info
> * The rest security description seems to go onto new lines without any tags. I think it shouldn't go onto new lines until the end of the line is reached.
> * Rest interaction code 'read' is used. However, do we also need 'vread' and 'search'? Will need to look at list of use cases if these are appropriate. From my interpreation of 'read', it looks like it's just to view only one resource, such as by id.
> * Is it needed to add the searchParams using hl7.org? I also can't seem to find these in the base standard but maybe in a place I didn't look. I would have thought we just need searchParams specific to the ODS use-case
> * Should the searchParam for ods-lastChangeDate use the type of 'date' instead of 'token'?

Publisher / contact, I suggest we feed that back in the team wide review / add to da - I don't consider them blockers for this

Agree re using the word conformance, changed

Jurisidction is a nightmare, GB-ENG is an iso 3166-2: https://en.wikipedia.org/wiki/ISO_3166-2:GB, but it's a bit of a weird one, and looks to have been added to the ios3166-2 after the last time FHIR refreshed its terminology, I think that'll probably need a special validation rule to ignore it (we have similar for loinc, and r4 languages), but agree we need to get Kev to clarify the system in the guidance doc for that

Agree with the removing version=current, changed

I'm working on the principle all NHSE stuff will be oauth at a minimum and we can adjust later 

No identified requirement for doing vread, but again, we can change later
 
HL7 search params are a weird one, in that you shouldn't need to specify them as they should be built in to most of the standard fhir servers, but we're essentially using them as a MustSupport flag on the server.
They are defined in a list on the bottom of the search param resource page: https://hl7.org/fhir/R4/searchparameter.html and in the registry https://hl7.org/fhir/R4/searchparameter-registry.html
 
Agree re: lastDatetime, changed

ryma2fhir commented on SearchParameters/England-SearchParameter-ODS-ActiveRole.xml:
```diff
@@ -0,0 +1,27 @@
+      <rank value="1" />
+    </telecom>
+  </contact>
+  <description value="SearchParameter for NHS England ODS FHIR API fopr querying OrganisationRole extension, roleCode sub element, where they are active" />
```
spelling mistake: fopr

APageNHS replied:
Good spot, fixed

ryma2fhir commented on SearchParameters/England-SearchParameter-ODS-Role.xml:
```diff
@@ -0,0 +1,28 @@
+  <base value="Organization" />
+  <type value="token" />
+  <expression value="Organization.extension('https://fhir.nhs.uk/England/StructureDefinition/Extension-England-OrganisationRole').extension('roleCode')" />
+  <xpath value="f:Organization/f:extension[@url='https://fhir.nhs.uk/England/StructureDefinition/Extension-England-OrganisationRole']/f:extension[@url='roleCode']" />
```
Is this needed as you have the expression. Also to be aware of from FHIR  

> Note: XPath is generally considered not useful because it does not apply to JSON and other formats and because of XSLT implementation issues, and may be removed in the future.

APageNHS replied:
Damn, thought i'd caught them all - removed, and also removed the xpathUsage as its a 0..1 and not needed if xpath is not present

ryma2fhir commented:
Received: "ERROR Organization.extension:extensionEnglandOrganisationRole: minimum required = 1, but only found 0 (from https://fhir.nhs.uk/England/StructureDefinition/England-Organization-ODS|1.0.0) [ Location - Organization] [ Location - Line 1, Col 1337]

ryma2fhir commented:
examples › ODS-Organization-4VE03-Example.xml › FHIR Validation: validation/src/common.js.ts#L62
expect(received).toBeFalsy()

Received: "ERROR Organization.extension:extensionEnglandOrganisationRole: minimum required = 1, but only found 0 (from https://fhir.nhs.uk/England/StructureDefinition/England-Organization-ODS|1.0.0) [ Location - Organization] [ Location - Line 1, Col 1758]
ERROR Organization.type: minimum required = 1, but only found 0 (from https://fhir.nhs.uk/England/StructureDefinition/England-Organization-ODS|1.0.0) [ Location - Organization] [ Location - Line 1, Col 1758]"

ryma2fhir commented:
examples › ODS-OrganizationAffiliation-G81086-Example.xml › FHIR Validation: validation/src/common.js.ts#L62
expect(received).toBeFalsy()

Received: "ERROR OrganizationAffiliation.identifier: minimum required = 1, but only found 0 (from https://fhir.nhs.uk/England/StructureDefinition/England-OrganizationAffiliation-ODS|1.0.0) [ Location - OrganizationAffiliation] [ Location - Line 1, Col 1764]
ERROR Validation failed for 'https://digital.nhs.uk/services/organisation-data-service/CodeSystem/ODSRelationship#RE4' [ Location - OrganizationAffiliation.code[0].coding[0]] [ Location - Line 1, Col 1760]"

digitalhealthmartin reviewed (commented):
- Perhaps the filename should be: CapabilityStatement-England-ODS-Requirements (because we may have different types of CapabilityStatements in the future for ODS and also would be inline with id etc)

digitalhealthmartin commented:
As for SearchParameters:

- Remove <?xml> tag at top as that was agreed at NHSE IG level

APageNHS commented:
Martins suggestions, plus general filename id/title/name fixes addressed
